lib/Arch.cpp

```diff
@@ -0,0 +1,441 @@
+  switch (id) {
+    case llvm::Type::VoidTyID: {
+      ret = "v";
+      break;
+    }
+    case llvm::Type::HalfTyID: {
+      ret = "e";
+      break;
+    }
+    case llvm::Type::FloatTyID: {
+      ret = "f";
+      break;
+    }
+    case llvm::Type::DoubleTyID: {
+      ret = "d";
+      break;
+    }
+    case llvm::Type::X86_FP80TyID: {
+      ret = "D";
+      break;
+    }
+    case llvm::Type::X86_MMXTyID: {
+      ret = "M";
+      break;
+    }
+    case llvm::Type::IntegerTyID: {
+      auto derived = llvm::cast<llvm::IntegerType>(type);
+      // Since there is no way to check for uint vs. int, lower all integer
+      // types to signed integers. Maybe come back when there is a better way to
+      // figure out unsigned vs signed ints
+      auto sign = true;
+      switch (derived.getBitWidth()) {
+        case 8: {
+          ret = sign ? "b" : "B";
+          break;
+        }
+        case 16: {
+          ret = sign ? "h" : "H";
+          break;
+        }
+        case 32: {
+          ret = sign ? "i" : "I";
+          break;
+        }
+        case 64: {
+          ret = sign ? "l" : "L";
+          break;
+        }
+      }
+      break;
+    }
+    case llvm::Type::FunctionTyID: {
+      ret = "func";
+      break;
+    }
+    case llvm::Type::StructTyID: {
+      auto struct_ptr = llvm::cast<llvm::StructType>(&type);
+      std::string element_list = "";
+      for (unsigned i = 0; i < struct_ptr->getNumElements(); i++) {
+        // TODO: BUG: this if statement protects against infinite chained
+        // pointers, which this code has a hard time with. Notably, this
+        // includes some types of C++ ostream functions.
+        if (struct_ptr->getElementType(i)->isPtrOrPtrVectorTy()) {
+          element_list += "?";
+          continue;
+        }
+        element_list += TranslateType(*struct_ptr->getElementType(i));
+      }
+      ret += " {" + element_list + "}";
+
+      break;
+    }
+    case llvm::Type::ArrayTyID: {
+      ret = "array";
+      break;
+    }
+    case llvm::Type::PointerTyID: {
+      ret = "*";
+      auto derived = llvm::dyn_cast<llvm::PointerType>(&type);
+      // Get the type of the pointee
+      ret += TranslateType(*derived->getElementType());
+      break;
+    }
+
+    default:
+      LOG(ERROR) << "Could not translate TypeID: " << id;
+      break;
+  }
+  return ret;
+}
+
+// Try to allocate a register for the argument based on the register constraints
+// and what has already been reserved. Return nullptr if there is no possible
+// register allocation.
+remill::Register *TryRegisterAllocate(
+    llvm::Type &type, std::vector<bool> &reserved,
+    const std::vector<RegisterConstraint> &register_constraints) {
+  SizeConstraint size_constraint;
+  TypeConstraint type_constraint;
+
+  if (type.isIntegerTy()) {
+    type_constraint = kTypeInt;
+    auto derived = llvm::cast<llvm::IntegerType>(type);
+    unsigned int width = derived.getBitWidth();
+    if (width == 64) {
```

Comment:
This needs to be made more general. I recommend using `llvm::DataLayout::fitsInLegalInteger` to figure out if we can likely fit it somewhere. Then instead of asking for the bit width of the type, ask for the type store size in bits or bytes, and match that to either 32 or 64.

Also, investigate what happens if you have a 128-bit integer passed as an argument or returned.